File: webmux/backend/src/api/agents.ts
import { Router, Request, Response, NextFunction } from 'express';
import { requireAuth, AuthPayload } from '../middleware/auth';
import { agentService } from '../services/agentService';
import { sessionBroker } from '../services/sessionBroker';
import { getAgentAccess } from '../services/agentAccess';

const router = Router();

function getOwner(req: Request): string {
  return (req as Request & { user?: AuthPayload }).user?.sub || 'anonymous';
}

function parseTermSize(body: { cols?: unknown; rows?: unknown }) {
  const cols = typeof body.cols === 'number' && Number.isFinite(body.cols) ? body.cols : 120;
  const rows = typeof body.rows === 'number' && Number.isFinite(body.rows) ? body.rows : 40;
  return {
    cols: Math.max(40, Math.min(240, Math.floor(cols))),
    rows: Math.max(10, Math.min(80, Math.floor(rows))),
  };
}

async function requireAgentAccess(_req: Request, res: Response, next: NextFunction): Promise<void> {
  try {
    const access = getAgentAccess();
    if (!access.allowed) {
      if (access.status !== 500) {
        await sessionBroker.deleteAgentSessions(access.error);
      }
      res.status(access.status).json({ error: access.error });
      return;
    }

    await sessionBroker.enforceAgentAccessPolicy();
    next();
  } catch (err) {
    console.error('Agent access cleanup failed:', err);
    res.status(500).json({ error: 'Failed to enforce agent access policy' });
  }
}

function sendInvalidAgent(res: Response): void {
  res.status(404).json({ error: 'Agent definition not found' });
}

router.use(requireAuth);

router.get('/config', (_req: Request, res: Response) => {
  try {
    res.json(agentService.getRuntimeConfig());
  } catch (err) {
    res.status(500).json({ error: (err as Error).message });
  }
});

router.get('/sessions', requireAgentAccess, async (_req: Request, res: Response) => {
  try {
    res.json(await agentService.listAllSessions());
  } catch (err) {
    console.error('Agent list error:', err);
    res.status(503).json({ error: 'Failed to list agent sessions' });
  }
});

router.get('/:agentId/sessions', requireAgentAccess, async (req: Request, res: Response) => {
  const config = agentService.getConfig(req.params.agentId);
  if (!config) {
    sendInvalidAgent(res);
    return;
  }

  try {
    res.json(await agentService.listSessions(config.id));
  } catch (err) {
    console.error(`${config.label} list error:`, err);
    res.status(503).json({ error: `Failed to list ${config.label} sessions` });
  }
});

router.post('/:agentId/attach', requireAgentAccess, async (req: Request, res: Response) => {
  const config = agentService.getConfig(req.params.agentId);
  if (!config) {
    sendInvalidAgent(res);
    return;
  }

  try {
    const { name } = req.body as { name?: string };
    if (!name || typeof name !== 'string') {
      res.status(400).json({ error: 'name is required' });
      return;
    }

    if (!(await agentService.hasSession(config.id, name))) {
      res.status(404).json({ error: `${config.label} session not found` });
      return;
    }

    const { cols, rows } = parseTermSize(req.body as { cols?: unknown; rows?: unknown });
    const result = await sessionBroker.ensureAgentAttach(
      getOwner(req),
      config.id,
      config.workspace,
      name,
      cols,
      rows,
      agentService.buildAttachExecArgv(config.id, name),
    );
    res.status(result.created ? 201 : 200).json(result.session);
  } catch (err) {
    console.error(`${config.label} attach error:`, err);
    res.status(500).json({ error: `Failed to attach ${config.label} session` });
  }
});

router.post('/:agentId/scratch', requireAgentAccess, async (req: Request, res: Response) => {
  const config = agentService.getConfig(req.params.agentId);
  if (!config) {
    sendInvalidAgent(res);
    return;
  }

  try {
    const { selectedName } = req.body as { selectedName?: string };
    if (selectedName !== undefined && typeof selectedName !== 'string') {
      res.status(400).json({ error: 'selectedName must be a string' });
      return;
    }
    const { cols, rows } = parseTermSize(req.body as { cols?: unknown; rows?: unknown });
    let cwd: string | undefined;
    if (selectedName) {
      if (!(await agentService.hasSession(config.id, selectedName))) {
        res.status(404).json({ error: `${config.label} session not found` });
        return;
      }
      cwd = await agentService.getPaneCurrentPath(config.id, selectedName);
    }
    const result = await sessionBroker.ensureAgentScratch(getOwner(req), config.id, config.workspace, cols, rows, cwd);
    res.status(result.created ? 201 : 200).json(result.session);
  } catch (err) {
    console.error(`${config.label} scratch error:`, err);
    res.status(500).json({ error: `Failed to create ${config.label} scratch shell` });
  }
});

export default router;

File: webmux/server/internal/httpapi/agents.go
package httpapi

import (
	"errors"
	"math"
	"net/http"

	"github.com/jordanhubbard/webmux/server/internal/agent"
)

func (s *Server) getAgentConfig(w http.ResponseWriter, r *http.Request, owner string) {
	value, err := s.sessions.Agents().Config()
	if err != nil {
		writeError(w, 500, err.Error())
		return
	}
	writeJSON(w, 200, value)
}

func (s *Server) agentAccess(w http.ResponseWriter, r *http.Request) (agent.Definition, bool) {
	service := s.sessions.Agents()
	c, err := service.Access("")
	// Enforce per-definition policy as well as the global enabled/user-count gate.
	// Config errors are deliberately non-destructive, matching the existing server.
	if cleanupErr := s.sessions.EnforceAgentAccess(); cleanupErr != nil {
		s.logger.Error("Agent access cleanup failed", "error", cleanupErr)
		writeError(w, 500, "Failed to enforce agent access policy")
		return agent.Definition{}, false
	}
	if err != nil {
		var access *agent.AccessError
		if errors.As(err, &access) {
			writeError(w, access.Status, access.Message)
		} else {
			writeError(w, 500, err.Error())
		}
		return agent.Definition{}, false
	}
	id := r.PathValue("agentId")
	if id == "" {
		return agent.Definition{}, true
	}
	d, ok := c.Find(id)
	if !ok {
		writeError(w, 404, "Agent definition not found")
	}
	return d, ok
}

func (s *Server) listAgents(w http.ResponseWriter, r *http.Request, owner string) {
	d, ok := s.agentAccess(w, r)
	if !ok {
		return
	}
	var items []agent.Session
	var err error
	message := "Failed to list agent sessions"
	if d.ID == "" {
		items, err = s.sessions.Agents().ListAll(r.Context())
	} else {
		message = "Failed to list " + d.Label + " sessions"
		items, err = s.sessions.Agents().List(r.Context(), d.ID)
	}
	if err != nil {
		s.logger.Warn(message, "error", err)
		writeError(w, 503, message)
		return
	}
	writeJSON(w, 200, items)
}

func agentTermSize(input map[string]any) (int, int) {
	dimension := func(key string, fallback, minimum, maximum int) int {
		value, ok := input[key].(float64)
		if !ok || math.IsNaN(value) || math.IsInf(value, 0) {
			return fallback
		}
		return int(math.Max(float64(minimum), math.Min(float64(maximum), math.Floor(value))))
	}
	return dimension("cols", 120, 40, 240), dimension("rows", 40, 10, 80)
}

func (s *Server) hasAgentSession(w http.ResponseWriter, r *http.Request, d agent.Definition, name, message string) bool {
	exists, err := s.sessions.Agents().HasSession(r.Context(), d.ID, name)
	if err != nil {
		s.logger.Warn(message, "error", err)
		writeError(w, 500, message)
		return false
	}
	if !exists {
		writeError(w, 404, d.Label+" session not found")
		return false
	}
	return true
}

func (s *Server) attachAgent(w http.ResponseWriter, r *http.Request, owner string) {
	d, ok := s.agentAccess(w, r)
	if !ok {
		return
	}
	var input map[string]any
	if !decodeJSON(w, r, &input) {
		return
	}
	name, ok := input["name"].(string)
	if !ok || name == "" {
		writeError(w, 400, "name is required")
		return
	}
	message := "Failed to attach " + d.Label + " session"
	if !s.hasAgentSession(w, r, d, name, message) {
		return
	}
	cols, rows := agentTermSize(input)
	value, created, err := s.sessions.EnsureAgentAttach(owner, d.ID, name, cols, rows)
	if err != nil {
		s.logger.Warn(message, "error", err)
		writeError(w, 500, message)
		return
	}
	status := 200
	if created {
		status = 201
	}
	writeJSON(w, status, value)
}
func (s *Server) scratchAgent(w http.ResponseWriter, r *http.Request, owner string) {
	d, ok := s.agentAccess(w, r)
	if !ok {
		return
	}
	var input map[string]any
	if !decodeJSON(w, r, &input) {
		return
	}
	name := ""
	if value, present := input["selectedName"]; present {
		var ok bool
		name, ok = value.(string)
		if !ok {
			writeError(w, 400, "selectedName must be a string")
			return
		}
	}
	message := "Failed to create " + d.Label + " scratch shell"
	cwd := ""
	if name != "" {
		if !s.hasAgentSession(w, r, d, name, message) {
			return
		}
		cwd = s.sessions.Agents().PaneCurrentPath(r.Context(), d.ID, name)
	}
	cols, rows := agentTermSize(input)
	value, created, err := s.sessions.EnsureAgentScratch(owner, d.ID, cols, rows, cwd)
	if err != nil {
		s.logger.Warn(message, "error", err)
		writeError(w, 500, message)
		return
	}
	status := 200
	if created {
		status = 201
	}
	writeJSON(w, status, value)
}
